lab/src/Microsoft.ServiceModel.Syndication/src/XmlReaderWrapper.cs

```diff
@@ -0,0 +1,575 @@
+﻿using System;
+using System.Threading.Tasks;
+using System.Xml;
+
+namespace Microsoft.ServiceModel.Syndication
+{
+    internal class XmlReaderWrapper : XmlReader
+    {
+        private XmlReader reader;
+
+        private Func<XmlReaderWrapper, Task<string>> getValueFunc;
+        private Func<XmlReaderWrapper, Task<string>> readElementStringFunc;
+        private Func<XmlReaderWrapper, Task<string>> readStringFunc;
+        private Func<XmlReaderWrapper, Task> readEndElementFunc;
+        private Func<XmlReaderWrapper, Task<XmlNodeType>> moveToContentFunc;
+        private Func<XmlReaderWrapper, string, string, Task> readStartElementFunc2;
+        private Func<XmlReaderWrapper, Task<bool>> isStartElementFunc;
+        private Func<XmlReaderWrapper, string, string, Task<bool>> isStartElementFunc2;
+        private Func<XmlReaderWrapper, Task> skipFunc;
+        private Func<XmlReaderWrapper, Task<bool>> readFunc;
+        private Func<XmlReaderWrapper, Task<string>> readInnerXmlFunc;
+
+        private XmlReaderWrapper(XmlReader reader)
+        {
+            if (reader == null)
+            {
+                throw new ArgumentNullException(nameof(reader));
+            }
+
+            this.reader = reader;
+
+            if (this.reader.Settings.Async)
+            {
+                InitAsync();
+            }
+            else
+            {
+                Init();
+            }
+        }
+
+        public static XmlReaderWrapper CreateFromReader(XmlReader reader)
+        {
```

Comment:
What we tried to do here is to get an XmlWriter, either with async support or not. The Wrapper is going to decide what to do depending on the capabilities of the writer.
I.E. If the writer does not support async, then it will execute sync methods.

As you can see in this method, if the reader is already wrapped it will just return the same reader without wrapping it.

---
In reply to: [126847030](https://github.com/dotnet/wcf/pull/2028#discussion_r126847030) [](ancestors = 126847030)